File: 407. Trapping Rain Water II.java
class Solution {
    public int trapRainWater(int[][] heightMap) {
        if (heightMap==null || heightMap.length==0 || heightMap[0].length==0) return 0;
        int row = heightMap.length; int col = heightMap[0].length;
        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> (a[0] - b[0]));
        HashSet<Integer> seen = new HashSet<>();
        int[] dx = {1, 0, 0, -1};
        int[] dy = {0, 1, -1, 0};
        int currmax = 0, ans = 0;

        for (int i=0; i<row; i++) {
            pq.offer(new int[] {heightMap[i][0], i, 0});
            seen.add(i*col);
            pq.offer(new int[] {heightMap[i][col-1], i, col-1});
            seen.add(i*col+col-1);
        }
        for (int j=1; j<col-1; j++) {
            pq.offer(new int[] {heightMap[0][j], 0, j});
            seen.add(j);
            pq.offer(new int[] {heightMap[row-1][j], row-1, j});
            seen.add((row-1)*col+j);
        }

        while (!pq.isEmpty()) {
            int[] curr = pq.poll();
            currmax = Math.max(curr[0], currmax);
            int r = curr[1], c = curr[2];
            for (int i=0; i<4; i++) {
                int nextRow = r + dx[i];
                int nextCol = c + dy[i];
                if (nextRow < 0 || nextRow >= row || nextCol < 0 || nextCol >= col) {
                    continue;
                }
                if (!seen.contains(nextRow*col+nextCol)) {
                    if (heightMap[nextRow][nextCol] < currmax) {
                        ans += currmax - heightMap[nextRow][nextCol];
                    }
                    pq.offer(new int[] {heightMap[nextRow][nextCol], nextRow, nextCol});
                    seen.add(nextRow*col+nextCol);
                }
            }
        }
        return ans;
    }
}

File: 407. Trapping Rain Water II.py
from heapq import heappush, heappop
class Solution:
    def trapRainWater(self, heightMap: List[List[int]]) -> int:
        if not heightMap or not heightMap[0]:
            return 0
        heap = []
        count = 0
        low = 0
        seen = set()
        m,n = len(heightMap), len(heightMap[0])
        for i in range(m):
            heappush(heap, (heightMap[i][0], i, 0))
            heappush(heap, (heightMap[i][n-1], i, n-1))
            seen.add(i*n)
            seen.add(i*n+n-1)
        for i in range(n):
            heappush(heap, (heightMap[0][i], 0, i))
            heappush(heap, (heightMap[m-1][i], m-1, i))
            seen.add(i)
            seen.add((m-1)*n+i)
        while heap:
            hei, row, col = heappop(heap)
            if hei > low:
                low = hei
            count += low - hei if hei < low else 0
            if row > 0 and (row-1)*n+col not in seen:
                heappush(heap, (heightMap[row-1][col], row-1, col))
                seen.add((row-1)*n+col)
            if row < m-1 and (row+1)*n+col not in seen:
                heappush(heap, (heightMap[row+1][col], row+1, col))
                seen.add((row+1)*n+col)
            if col > 0 and row*n+col-1 not in seen:
                heappush(heap, (heightMap[row][col-1], row, col-1))
                seen.add(row*n+col-1)
            if col < n-1 and row*n+col+1 not in seen:
                heappush(heap, (heightMap[row][col+1], row, col+1))
                seen.add(row*n+col+1)
        return count